README.md
GW Logicat 
==========
[TOC]

## CHANGELOG

## Intro 
Main doc about **logicat-iot** project. This is an iot gateway developeded (v1.0.0 on 20/05/2020) on snap, node.js v13.14.0 where data are gather from the field and sent to GCP ecosystem. Firebase real-time database, Firestore, GCP Pub/Sub pattern, GCP Cloud Functions, BigQuery are involved! Iot device can handle the device data and the field data into separate cloud connections with two different security cahnnels one fot the device status and configuration and the other for client's data.

Actually (22-09-2021) Running on:

- Snap
- Node.js v16.9.1
- Redis v6.2.5
- Firebase V9 API
- @google-cloud/pubsub v2.17.0

## Getting Start 
The OS in charge is Ubuntu Core, starting from 04/2020 with the Ubuntu Core 20-04 LTS version, power on and configure the Ubuntu Core system with the basic network configurations and login in the right Ubuntu ONE account. The Github code has a snapcraft.yaml file to build the snap and install on all the Linux architecture you need to run the code.

Actually (22-09-2021) Running on:

- base: core20
- architecture: amd64, arm64, armhf

## DEVICE_SERIAL 
Device serial is the unique serial device identification on the Cloud. The serial is bound to the unique id based on snapd installation.To provision new device you need the Device serial due to register the new device on the Firebase device management cloud service. Sometimes when a device is swap with a new device (and new snap serial) the old device snap serial is portable to the new one so there is a difference between the device snap serial and the DEVICE_SERIAL env. **Keep in mind that the DEVICE_SERIAL is the serial that identify the device on the cloud** so check the DEVICE_SERIAL value before use the snap serial. 
```bash
# see env
echo $DEVICE_SERIAL

# catch the snap serial
snap model | grep serial 
```

## Cloud Components
Here the components involved: 
+ Google Cloud aervice-accounts
+ Google Cloud Pub/Sub topics
+ Firebase Cloud Functions
+ Firebase Real-time db  (conf, status)
+ Firebase Firestore db  (device info, webapp data)

## Service-Account
The authentication & authorization to push the data to the cloud (PubSub client) is handle with the service-accounts permissions installed on every device. Every client's group of devices have the same service-account json security file copied and linked with an env variable. Here the service-accounts available: 
```bash
gcloud beta iam service-accounts list
> ... 
```

Of course you need to make the env variable PERSISTENT, so create the right file in the **/etc/profile.d** folder: Here an example: 

```bash
# google_application_credentials.sh  file  /etc/profile.d 
export GOOGLE_APPLICATION_CREDENTIALS="/var/snap/<snap>/common/keys.json"

# ubuntu core /etc/environment 
sudo vi /etc/environment 
# GOOGLE_APPLICATION_CREDENTIALS="/var/snap/<snap>/common/keys.json"

# load the new env variable without re-loggin: 
source /etc/profile.d/google_application_credentials.sh
``` 

In Ubunto Core you can set the **/etc/environment** file to add the env you need to run the app.

## env 
Here the list of env needed to authenticate to the Cloud services: 

DEVICE_SERIAL
GOOGLE_APPLICATION_CREDENTIALS
FIREBASE_CONF
FIREBASE_PSW
FIREBASE_EMAIL

## JSON Messages 
The fields PLC, Robots, Sensors and all other TCP comm can ship data to iot gateway. The device do a control on JSON message before push to the cloud, here the steps to validate the message: 

+ JSON Formatted message 
+ id field, a valid sending device id
+ t field, in each message the t field needed (empty string or valid timestamp)

## initPubSub 
Every time the communication with the Pub/Sub service is activated a message is sent, this message is published to Pub/Sub topic, ready to be insert into serial id events table: 

```javascript
{"id":"${deviceId.replace(/-/g,'_')}", "t":${Math.floor(Date.now() / 1000)}, "msg": "topic:${formattedTopic}"}
```

Table eg. 

project-id.table-id.serial-id_a40f_4x44_..... 

## Redis
The heart of data saving is a Redis instance, to configure / stop / restart the service follow the commands:
```bash
# check the service
snap services

# configure redis
vim /var/snap/logicat-iot/common
```



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [node16 branch] - xx-2021
### Update
  - Node.js from 14.17.6 to 16.9.1

## [2.7.0] - 10-10-2021
### Change
  - fix the deviceStatusRef.set old function
  - minor fix

## [2.6.2] - 24-09-2021
### Removed
  - architecture armhf

## [2.6.1] - 22-09-2021
### Change
  - Firebase v9 API
  - minor fix

## [2.5.0] - 20-09-2021
### change
  - version to 2.5.0

## [2.4.0] - 20-09-2021
### Change
  - Minor fix for compile the snapcraft.yaml
  - back to stable line

## [2.1.0] - 20-09-2021
### Change
  - devel in snapcraft.yaml

## [2.0.0] - 14-09-2021
### Change
  - Ubuntu Core 20
  - Logicat icon
  - snapcraft.yaml minor details
  - redis part of snapcraft.yaml

## Update
  - @google-cloud/pubsub from 2.5.0 to 2.17.0
  - firebase from 7.22.1 to 9.0.2
  - ioredis from 4.17.3 to 4.27.9
  - public-ip from 4.0.2 to 4.0.4

## [1.4.1] - 2020-01-07
### Change
  - Minor code style fix

## [1.4.0] - 2020-10-06
### Change 
  - Fix Error: Retry total timeout exceeded before any response was received in PubSub 
  - Raise maxWaitTime from 10 to 60 
  - Update the @google-cloud/pubsub to ^2.5.0 
  - Update the firebase to ^7.22.1
  - Update the ioredis to ^4.17.3
  - Update Node.js to v14.13.0

## [1.3.0] - 2020-07-08
### Changed
  - Fix the table.serial-id - to _ pattern 

## [1.2.0] - 2020-06-15
### Changed
  - Fix the publishBatchedMessages

## [1.1.0] - 2020-0-12
### Added
  - Control to undefined to fireUser and batchPublisher
  - @ERROR messages  
  - deviceId with _ for BigQuery table 

### Changed 
  - Redis v6.0.5

## [1.0.8] - 2020-06-03
### Added 
  - logs details into in-cache db 
  - errors details into in-cache db
  - more details on init device msg

### Changed 
  - snap - remove usless home permission  

## [1.0.7] - 2020-05-25
### Added 
  - Bootstrap the Project by [@cicciosgamino](https://github.com/CICCIOSGAMINO)

index.mjs
#!/usr/bin/env node

'use strict'
// ip, interface handling
import {
  internalIp4,
  activeIface,
  publicIp4
} from './network.mjs'
// DNS Connection module
import {
  connectionEmitter,
  startConnectionChecking,
  isInternetAvailable
} from './connection.mjs'
// Ids Validator
import {
  checkId
} from './validator.mjs'

// ---------------------------------------------- Redis -------------------------------------------
// Redis in-cache db is responsable to store/retrive iot items and offline/online configurations
import {
  redisSaveConfiguration,
  redisGetConfiguration,
  redisAddIotItem,
  redisCountIotItems,
  redisGetIotBatchItems,
  redisAddLog,
  redisAddErr
} from './redis.mjs'

// https://firebase.google.com/docs/database/web/read-and-write
import { initializeApp } from 'firebase/app'
import { getAuth, signInWithEmailAndPassword, onAuthStateChanged } from 'firebase/auth'
import { getDatabase, ref, set, onValue, serverTimestamp, onDisconnect } from 'firebase/database'

// Google Cloud PubSub
import pubsub from '@google-cloud/pubsub'
// net (TCP Server)
import { Server } from 'net'
const server = new Server()

// Creates a client; cache this for further use
const pubSubClient = new pubsub.PubSub()
// const publisherClient = new pubsub.v1.PublisherClient({})

// Google Pub/Sub formatted Topic projects/[project_id]/topics/[topic_name]
// const formattedTopic = publisherClient.topicPath(projectId, topicName);
let formattedTopic, batchPublisher
// maxMessages and maxWaitTime [sec]
const maxMessages = 1
const maxWaitTime = 60

// Device Serial (Unique by OS eg Snap serial, hardware serial ...)
const deviceId = process.env.DEVICE_SERIAL || ''

// Default PORT and INTERVAL
const DEFAULT_PORT = 8888
const DEFAULT_INTERVAL = 36000
const DEFAULT_DEVICES_IDS = []
// Array with ids list of devices can push messages
let devicesIds = DEFAULT_DEVICES_IDS

// -------------------------------------------- Connection ----------------------------------------
// Cheking the connection status with polling Google DNS, handled with the connectionEmitter!

connectionEmitter.on('connected', (data) => {
  // Timing + Connection OK
  log(`@SCHEDULED (Online): ${data.interval}sec`)

  // Init services
  initPubSub()
  fireLogin(email, passw)

  // if Auth and PubSub active do stuff
  if (fireUser !== undefined && batchPublisher !== undefined) {
  // logged, push items of zero is returned
    redisCountIotItems()
      .then(count => {
      // if no iot items to push return
        if (count > 0) {
          redisGetIotBatchItems(100)
            .then(items => {
              items.forEach(i => {
                // console.log(`@I (${typeof i}): ${i}`)
                // publish the messages
                publishBatchedMessages(i)
                  .then(messageId => {
                    log(`@MSG (Published): ${messageId}`)
                  })
                  .catch(err => {
                    // ERROR in Publish - CONSOLE
                    log(`@ERROR (PUB-ITEM): ${err}`)
                  })
              })
            })
            // catch on redisCountIotItems
            .catch(err => {
              console.log(`@ERROR (REDIS_ITEMS) Get IoT items: ${err}`)
            })
        } // end if count > 0
      })
      .catch(err => {
        log(`@ERROR (REDIS_COUNT) Count Iot items: ${err}`)
      })
  }
})

connectionEmitter.on('disconnected', (data) => {
  // Timing but NOT Online
  log(`@SCHEDULED (Offline): ${data.interval}sec`)
})

// If Offline && Configuration present active LOCAL TCP services
isInternetAvailable()
  .then(_ => {
    // Internet Available - Fetch from Firebase the config
    fireLogin(email, passw)
  })
  .catch(_ => {
    // Internet NOT Available - Fetch from Redis db if Config present
    redisGetConfiguration()
      .then(configObj => {
        // Security check on port and interval
        const port = Number(configObj.port) || DEFAULT_PORT
        const interval = Number(configObj.interval) || DEFAULT_INTERVAL
        devicesIds = configObj.ids || DEFAULT_DEVICES_IDS
        // Init the TCP Server
        initTCPServer(port)

        // Init scheduler with Offline config
        startConnectionChecking(interval)
      })
      .catch(err => {
        log(`@ERROR (Redis): Offline Configurations ${err}`)
      }) // End Catch Redis Conf
  }) // End Catch isInternetAvailable

// --------------------------------------------- Firebase -----------------------------------------
// Real-time db is the service to handle the device's configuration via cloud.(topics, timing, etc)
let fireUser

// Load ENV VARS for Firebase
const fireConf = JSON.parse(process.env.FIREBASE_CONF)
const email = process.env.FIREBASE_EMAIL || ''
const passw = process.env.FIREBASE_PSW || ''

// Firebase Init
initializeApp(fireConf)
const auth = getAuth()
const db = getDatabase()

// Firebase References
const fireStatusRef = ref(db, '.info/connected')
const refConfig = ref(db, `/devices/${deviceId}/config`)
const deviceStatusRef = ref(db, `/devices/${deviceId}/status`)
const errorRef = ref(db, `/devices/${deviceId}/errors`)

/**
 * Firebase signin - Try to login to Firebase Cloud Infrastructure
 *
 */
const fireLogin = (email, passw) => {
  // check auth
  if (fireUser !== undefined) return

  // auth in firebase
  signInWithEmailAndPassword(auth, email, passw)
    .catch(err => {
      // ERROR in LOGIN - CONSOLE
      log(`@ERROR (Firebase_Auth): ${err.message}`, 1)
    })
}

// Listening for Firebase Auth
onAuthStateChanged(auth, user => {
  if (user) {
    // LOGGED - CONSOLE
    fireUser = user.uid
    log(`@USER (Logged): ${user.uid}`)

    startStatusListening()
    startListenerConfig()
  } else {
    fireUser = undefined
  }
})

/**
  * Start the Listening for Firebase Config object needed for start PubSub and TCP
  */
const startListenerConfig = () => {
  // create the firebase configChanged event handler
  const configChanged = function (snap) {
    // save the Firebase configuration object in Redis
    if (snap.val()) {
      redisSaveConfiguration(snap.val())
    }

    redisGetConfiguration()
      .then(configObj => {
        const port = Number(configObj.port) || DEFAULT_PORT
        const interval = Number(configObj.interval) || DEFAULT_INTERVAL
        devicesIds = configObj.ids || DEFAULT_DEVICES_IDS

        formattedTopic = configObj.topic || ''
        log(`@MSG (Config): Port: ${port} Interval: ${interval} Topic: ${formattedTopic} `)

        // Init the Timing
        startConnectionChecking(interval)

        // Init the PubSub Service
        initPubSub()
          .catch(error => {
            // ERROR initPubSub - CONSOLE
            log(`@ERROR (PubSub Init): ${error}`, 2)
          })
        // Init the TCP Server
        initTCPServer(port)
          .catch(error => {
            // ERROR initTCPServer - CONSOLE
            log(`@ERROR (TCP Init): ${error}`, 2)
          })
      })
      .catch(err => {
        // catch the JSON.parse or Redis get exceptions
        log(`@ERROR (JSON Parser or Redis Exec): ${err}`, 2)
      })
  }

  const errorConfig = function (error) {
    // ERROR in CONFIG - CONSOLE
    // 1) Bad DEVICE_SERIAL
    // 2) No Connection
    log(`@ERROR (Firebase_Conf): ${error}`, 2)
  }

  onValue(refConfig, configChanged, errorConfig)
}

/**
  * Firebase ONLINE/OFFLINE status
  *
  */
const startStatusListening = () => {
  // status of firebase connection
  onValue(fireStatusRef, async snap => {
    // snap.val()  true/false
    if (snap.val()) {
      set(deviceStatusRef, {
        t: serverTimestamp(),
        iface: activeIface(),
        ip: internalIp4(),
        public: await publicIp4()
      })
    } else {
      onDisconnect(deviceStatusRef).set({
        t: serverTimestamp(),
        iface: '',
        ip: '',
        public: ''
      })
    }
  })
}

// -------------------------------------------- PubSub --------------------------------------------
// Google Cloud PubSub is the connector with the cloud, topics are the highway to drive the data!

/**
  * Init the Google Cloud PubSub service with the topic received by Config obj
  *
  */
const initPubSub = async () => {
  // check the publisher
  if (batchPublisher !== undefined) return

  // Init the Google PubSub service and test topic permission
  const pushMsgPermission = ['pubsub.topics.publish']
  pubSubClient
    .topic(formattedTopic)
    .iam.testPermissions(pushMsgPermission)
    .then(p => {
      // Permission OK (can push on this topic )
      // Init the batchPublisher
      batchPublisher = pubSubClient.topic(formattedTopic, {
        batching: {
          maxMessages: maxMessages,
          maxMilliseconds: maxWaitTime * 1000
        }
      })

      // Register the event to status device - Every time initPubSub send this formatted service message
      // This is very useful to test the PubSub / Cloud Function service
      publishBatchedMessages(`
        {"id":"${deviceId.replace(/-/g, '_')}", "t":${Math.floor(Date.now() / 1000)}, "msg": "topic:${formattedTopic}"}
      `).then(messageId => {
        log(`@SERVICE (INIT_PUBSUB): ${messageId}`)
      })
    })
    .catch(err => {
      log(`@ERROR (PubSub): Init Service: ${err}`, 1)
    })
}

/**
 * Publish to PubSub topic with the retry params set
 *
 * @param { string } msg - JSON formatted message to Publish
 */
// const publishWithRetrySettings = async (msg) => {
//   const dataBuffer = Buffer.from(msg)
//   const messagesElement = {
//     data: dataBuffer
//   }
//   const messages = [messagesElement]
//   const request = {
//     topic: formattedTopic,
//     messages: messages
//   }

//   // How the publisher handles retryable failures
//   const retrySettings = {
//     retryCodes: [
//       10, // 'ABORTED'
//       1, // 'CANCELLED',
//       4, // 'DEADLINE_EXCEEDED'
//       13, // 'INTERNAL'
//       8, // 'RESOURCE_EXHAUSTED'
//       14, // 'UNAVAILABLE'
//       2 // 'UNKNOWN'
//     ],
//     backoffSettings: {
//       initialRetryDelayMillis: 1000,
//       retryDelayMultiplier: 1.3,
//       maxRetryDelayMillis: 60000,
//       initialRpcTimeoutMillis: 5000,
//       rpcTimeoutMultiplier: 1.0,
//       maxRpcTimeoutMillis: 600000,
//       totalTimeoutMillis: 6000000
//     }
//   }

//   const [response] = await publisherClient.publish(request, {
//     retry: retrySettings
//   })
//   log(`@MSG (Published): ${response.messageIds}`)
// }

/**
 * Publish to PubSub topic in batch mode
 *
 * @param {String} msg - JSON formatted message to Publish or [{...}, {...}, ...] for messages
 */
const publishBatchedMessages = (msg) => {
  const dataBuffer = Buffer.from(`${msg}`)
  return batchPublisher.publish(dataBuffer)
}

// ----------------------------------------------- Log --------------------------------------------
// The log, errors and exception can be logged into firebase(connected/auth), redis and console!

/**
 * Function to log messages in console and redis (firebase only for severe errors)
 *
 * @param {string } msg - message to log
 * @param { number } severity - level of severity 0 log, 1 error, 2 error cloud log
 */
const log = (msg, severity = 0) => {
  switch (severity) {
    // 0  console & redis log
    case 0:
      console.log(msg)
      // write in redis log
      redisAddLog(msg)
        .then(_ => 'ok')
        .catch(err => {
          console.log(`@ERROR (Redis) Writing Log: ${err}`)
        })
      break
    // 1 console & redis error log
    case 1:
      console.log(msg)
      // write in redis err
      redisAddErr(msg)
        .then(_ => 'ok')
        .catch(err => {
          console.log(`@ERROR (Redis) Writing Err: ${err}`)
        })
      break
    // 2 console & redis error log & firebase error log
    case 2:
      console.log(msg)
      // write in redis err
      redisAddErr(msg)
        .then(_ => 'ok')
        .catch(err => {
          console.log(`@ERROR (Redis) Writing Err: ${err}`)
        })
      // write in firebase error
      set(errorRef, {
        t: serverTimestamp(),
        error: msg
      })
        .catch(err => {
          console.log(`@ERROR (Firebase) Writing Err: ${err}`)
        })
      break
  }
}

// -------------------------------------------- TCP Server ----------------------------------------
// TCP Server is in charge to hadnle the local TCP communication with the devices on the field!

/**
  * Init the TCP Server configure the port received by Config obj
  *
  * @param {Number} port - TCP Server port
  *
  */
const initTCPServer = async (port) => {
  port = Number(port)
  // close actual server
  server.close()

  // Active and listen on new port
  server.listen(port, () => {
    log(`@SERVER (${port}): Listening`)
  })
}

server.on('connection', (socket) => {
  // Data received from Client
  socket.on('data', (chunk) => {
    // Test is a RIGHT JSON object, only JSON msg can be handled
    try {
      // needed because can't use double quote in ABB Robot sw
      const rightApexChunk = chunk.toString().replace(/'/g, '"')
      const data = JSON.parse(rightApexChunk, (key, value) => {
        // checking the t field, must contain valid timestamp
        if (key === 't' && value === '') {
          return Math.floor(Date.now() / 1000)
        } else {
          return value
        }
      })

      // Alone iot item in Object form
      if (data.length === undefined) {
        // only one JSON object to store
        if (checkId(data, devicesIds)) {
          redisAddIotItem(data)
            .then(record => {
              // ok
              socket.write('1')
              log(`@RECORD (${record}): ${JSON.stringify(data)}`)
            })
            .catch(err => {
              // error on redis add item
              log(`@ERROR(ID) ${err}`)
              socket.write('0')
            })
        } else {
          // not correct id of one item
          socket.write('0')
          log('@ERROR Not correct id of one iot item', 1)
        }
      }
      // More than one iot item in Array form
      if (data.length > 0) {
        // check if all ids are correct
        let correct = true
        data.forEach(element => {
          if (!checkId(element, devicesIds)) {
            correct = false
          }
        })

        // if all ids correct add to redis
        if (correct) {
          const promises = []

          data.forEach(element => {
            promises.push(redisAddIotItem(element))
          })

          Promise.all(promises)
            .then(responses => {
              // OK
              socket.write('1')
              log(`@REDIS (Multy Insert): ${responses}`)
            })
            .catch(err => {
              // ERR
              socket.write('0')
              log(`@ERROR (Redis): ${err}`, 1)
            })
        } else {
          // not correct id of one item
          socket.write('0')
          log('@ERROR (Multy ids): Not correct id of one iot item', 1)
        }
      }
    } catch (e) {
      // ERROR in JSON parsing - CONSOLE
      socket.write('0')
      log(`@ERROR (JSON Parsing): ${e}`, 2)
    }
  }) // End Socket.on('data', chunck => .... )

  // Connection
  socket.on('connect', () => {
    console.log(`@SOCKET (Connect): ${socket.remoteAddress}:${socket.remotePort}`)
  })

  // Close
  socket.on('close', () => {
    console.log(`@SOCKET (Close): ${socket.remoteAddress}:${socket.remotePort}`)
  })

  // Client request to END the TCP connection, server END the connection
  socket.on('end', () => {
    console.log(`@SOCKET (End): ${socket.remoteAddress}:${socket.remotePort}`)
  })

  // Don't forget to catch error, for your own sake.
  socket.on('error', (err) => {
    // ERROR on Socket - CONSOLE
    log(`@ERROR (Socket): ${err}`, 1)
  })
}) // End server.connection

// ------------------------------------------ Main Process ----------------------------------------
// Main process handling and uncaught exceptions (all exceptions that are not handled in the code )

// warnings
process.on('warning', (warning) => {
  log(`@WARNING (Process): ${warning.name} - ${warning.message}`, 1)
})

// emitted whenever a Promise is rejected and no errors handler is attached
process.on('unhandledRejection', (reason, promise) => {
  log(`@UNHANDLED (Rejection): ${promise} - ${reason}`, 2)
})

// event is emitted before an 'uncaughtException'
process.on('uncaughtException', (err, origin) => {
  log(`@UNCAUGH (Exception): ${origin} - ${err}`, 2)
})


network.mjs
// Utility to get Network Interaface details
import { networkInterfaces } from 'os'
import publicIp from 'public-ip'

/**
  * Filter from networkInterfaces the external IPv4 of the Active Interface
  *
  * @return { string }   - The IP Address IPv4
  */
export const internalIp4 = () => [].concat.apply([], Object.values(networkInterfaces()))
  .filter(details => details.family === 'IPv4' && !details.internal)
  .pop().address

/**
  * Filter from networkInterfaces the Active with the loopback interface NOT included
  *
  * @return { string }   - Active interface
  */
export const activeIface = () => [].concat.apply([], Object.keys(networkInterfaces()))
  .filter(i => i !== 'lo')
  .pop()

/**
  * Get the public ip on internet IPv4
  *
  * @return { string }   - Return the Public Ipv4 address
  */
export const publicIp4 = async () => {
  return await publicIp.v4()
}


connection.mjs
// Utility to check the Internet connection throught DNS polling
import { lookupService } from 'dns'
import { EventEmitter } from 'events'

// Min & default updateing time interval (seconds)
const minUpdatingInterval = 5
const defaultUpdatingInterval = 36000

// Google DNS (used to check internet)
const GOOGLE_DNS_SERVER_IP = '8.8.8.8'
const GOOGLE_DNS_SERVER_PORT = 53

let checkConnectionTimer
let interval

class ConnectionEmitter extends EventEmitter {}

export const connectionEmitter = new ConnectionEmitter()

/*
connectionEmitter.on('connected', (data) => {
  // Connected to Internet   data.hostname  data.service
  ...
})

connectionEmitter.on('disconnected', (data) => {
  // Disconnected from Internet
  ...
})

connectionEmitter.on('errors', (err) => {
  // Handle Errors with Emitter
  ...
}) */

/**
 * Check Internet Connection is NOT availability
 *
 * @returns { Promise } connection NOT available
 */
export const isInternetAvailable = () => {
  return new Promise((resolve, reject) => {
    // dns lookup service on Google DNS Server
    lookupService(GOOGLE_DNS_SERVER_IP, GOOGLE_DNS_SERVER_PORT,
      (err, hostname, service) => {
        if (err) {
          // NO Internet Available
          reject(new Error('Internet NOT available'))
        } else {
          // Internet Available
          resolve('Internet available')
        }
      })
  })
}

/**
 * service based on Google DNS service to check connection's status
 */
const connectionLookupService = () => {
  // dns lookup service on Google DNS Server
  lookupService(GOOGLE_DNS_SERVER_IP, GOOGLE_DNS_SERVER_PORT,
    (err, hostname, service) => {
      if (err) {
        connectionEmitter.emit('disconnected', {
          interval: interval
        })
      } else {
        connectionEmitter.emit('connected', {
          hostname: hostname,
          service: service,
          interval: interval
        })
      }
    })
}

/**
 * Start the connection checking service, no arguments start with default interval
 *
 * @param { number | null } [ seconds ] - Time Interval in seconds > 1
 */
export const startConnectionChecking =
  (seconds = defaultUpdatingInterval) => {
    // valid number for seconds in interval check
    if (typeof seconds === 'number' && seconds >= minUpdatingInterval) {
      const intervalInMillis = seconds * 1000
      clearInterval(checkConnectionTimer)
      checkConnectionTimer =
        setInterval(connectionLookupService, intervalInMillis)
      interval = seconds
    } else {
      // not valid argument passed
      connectionEmitter.emit('error',
        'Not valid time interval value passed, restored default')
      startConnectionChecking(defaultUpdatingInterval)
      interval = defaultUpdatingInterval
    }
  }

/**
 * Stop the connection checking service
 */
export const stopConnectionCheking = () => {
  clearInterval(checkConnectionTimer)
}


index.js
#!/usr/bin/env node

'use strict';
// net (TCP Server)
const net = require('net');
const server = new net.Server();

// fire.js - Prototype of  main.js 
const { internalIp4, activeIface, publicIp4 } = require('./utils.cjs');

// Firebase module 
const firebase = require('firebase/app');
require("firebase/auth");
require("firebase/database");

// Google Cloud PubSub  
const {PubSub} = require('@google-cloud/pubsub');
// Creates a client; cache this for further use
const pubSubClient = new PubSub();

// Creates a publisher client 
const {v1} = require('@google-cloud/pubsub');
const publisherClient = new v1.PublisherClient({});

// Google Pub/Sub formatted Topic projects/[project_id]/topics/[topic_name]
// const formattedTopic = publisherClient.topicPath(projectId, topicName);
let formattedTopic;
let batchPublisher;
// maxMessages and maxWaitTime [sec]
const maxMessages = 10;
const maxWaitTime = 60;

// Retrive the Device Serial (Unique by OS eg Snap serial, hardware serial ...)
const deviceId = process.env['DEVICE_SERIAL'] || '';

// ----------------------------------------- Firebase -----------------------------------
// Load ENV VARS for Firebase
const fireConf = JSON.parse(process.env['FIREBASE_CONF']);
const email = process.env['FIREBASE_EMAIL'] || '';
const passw = process.env['FIREBASE_PSW'] || '';
 
// Firebase Init 
const app = firebase.initializeApp(fireConf);
const auth = app.auth();
const db = app.database();

// Firebase References 
const fireStatusRef = db.ref('.info/connected');
const refConfig = db.ref(`/devices/${deviceId}/config`);
const deviceStatusRef = db.ref(`/devices/${deviceId}/status`);
const errorRef = db.ref(`/devices/${deviceId}/errors`);

// Firebase Auth the User 
auth.signInWithEmailAndPassword(email, passw)
    .catch(err => {
      // ERROR in LOGIN - CONSOLE 
      console.log(`@ERROR(FIREBASE_AUTH): ${err.message}`);
    }); 

auth.onAuthStateChanged(user => {
  if(user) {
    // LOGGED - CONSOLE 
    console.log(`@USER: ${ user.uid }`);

      startStatusListening();
      startListenerConfig();

  }
}); 

/**
  * Start the Listening for Firebase Config object needed for start PubSub and TCP
  * @return { }
  */
function startListenerConfig() {

  // create the firebase configChanged event handler 
  const configChanged = function(snap) {

    const configObj = snap.val();
    const port = Number(configObj.port) || 8888;
    formattedTopic = configObj.topic || "";
    console.log(`@CONFG: Port: ${port} Topic: ${formattedTopic} `);

    // Init the PubSub Service 
    initPubSub()
      .catch(error => {
        // ERROR initPubSub - CONSOLE 
        const msg = `@ERROR(PubSub): ${error}`;
        console.log(msg);
        logError(msg);
      });
    // Init the TCP Server 
    initTCPServer(port)
      .catch(error => {
        // ERROR initTCPServer - CONSOLE 
        const msg = `@ERROR(TCP): ${error}`;
        console.log(msg);
        logError(msg);
      })
  }

  const errorConfig = function(error) {
    // ERROR in CONFIG - CONSOLE 
    // 1) Bad DEVICE_SERIAL 
    // 2) No Connection 
    const msg = `@ERROR(FIREBASE_CONF): ${error}`;
    console.log(msg);
    logError(msg);
  }

  refConfig.on('value',configChanged, errorConfig);
}


/**
  * Start the status ONLINE/OFFLINE device listening on firebase (only after firebase login)
  * @return { }
  */
function startStatusListening() {

  // Listening for the status of firebase connection 
  fireStatusRef.on(`value`, async snap => {
    // snap.val()  true/false  
    if(snap.val()){
      deviceStatusRef.set({
        t: firebase.database.ServerValue.TIMESTAMP,
        iface: activeIface(),
        ip: internalIp4(),
        public: await publicIp4()
      });
    } else {
      deviceStatusRef.onDisconnect().set({
        t: firebase.database.ServerValue.TIMESTAMP,
        iface: '',
        ip: '',
        public: ''
      });
    }
  });
};

/**
 * If connected log the main error on Firebase db 
 * @param {String} msg - Error message to log in the Firebase db 
 * @return {Promise} 
 */
function logError(err) {
  return errorRef.set({
    t: firebase.database.ServerValue.TIMESTAMP,
    error: err
  });
};

// -------------------------------------------- PubSub -------------------------------------------

/**
  * Init the Google Cloud PubSub service with the topic received by Config obj
  * @return { }
  */
async function initPubSub() {

    // Init the Google PubSub service and test topic permission 
    const pushMsgPermission = ['pubsub.topics.publish'];
    const permission = await pubSubClient
      .topic(formattedTopic)
      .iam.testPermissions(pushMsgPermission);

      
    // Init the batchPublisher 
    batchPublisher = pubSubClient.topic(formattedTopic, {
      batching: {
        maxMessages: maxMessages,
        maxMilliseconds: maxWaitTime * 1000
      }
    });

    // Send Msg every time PubSub settings change 
    publishBatchedMessages(`
        {"id":"0", "t":${Math.floor(Date.now() / 1000)}, "msg": "topic:${formattedTopic}"}
    `)
};

/**
 * Publish to PubSub topic with the retry params set 
 * 
 * @param {String} msg - JSON formatted message to Publish 
 */
async function publishWithRetrySettings(msg) {

  const dataBuffer = Buffer.from(msg);
  const messagesElement = {
    data: dataBuffer,
  };
  const messages = [messagesElement];
  const request = {
    topic: formattedTopic,
    messages: messages
  };

  // How the publisher handles retryable failures 
  const retrySettings = {
    retryCodes: [
      10, // 'ABORTED'
      1, // 'CANCELLED',
      4, // 'DEADLINE_EXCEEDED'
      13, // 'INTERNAL'
      8, // 'RESOURCE_EXHAUSTED'
      14, // 'UNAVAILABLE'
      2, // 'UNKNOWN'
    ],
    backoffSettings: {
      initialRetryDelayMillis: 1000,
      retryDelayMultiplier: 1.3,
      maxRetryDelayMillis: 60000,
      initialRpcTimeoutMillis: 5000,
      rpcTimeoutMultiplier: 1.0,
      maxRpcTimeoutMillis: 600000,
      totalTimeoutMillis: 6000000,
    },
  };

  const [response] = await publisherClient.publish(request, {
      retry: retrySettings,
    });

  console.log(`@MSG: ${response.messageIds}`)

}

/**
 * Publish to PubSub topic in bath mode
 * 
 * @param {String} msg - JSON formatted message to Publish or [{...}, {...}, ...] for messages
 */
async function publishBatchedMessages(msg) {
  const dataBuffer = Buffer.from(msg);

  const messageId = await batchPublisher.publish(dataBuffer);
  // const messageId = await batchPublisher.publishJSON(dataBuffer);
  console.log(`@MSG: ${messageId} Published!`);

}

// ------------------------------------------- TCP Server ----------------------------------------

/**
  * Init the TCP Server configure the port received by Config obj
  * @param {Number} port - TCP Server port 
  * @return { }
  */
 async function initTCPServer(port) {
    port = Number(port);

    // close actual server 
    server.close();

    // Active and listen on new port 
    server.listen(port, () => {
      console.log(`@SERVER: Active(port:${port})`);
    });
 };

 server.on('connection', (socket) => {

   // TCP connection established 
   socket.write(`CONNECTED`);

   //Data received from Client 
   socket.on('data', (chunk) => {

    // Test is a RIGHT JSON object, ready to Dataflow template, and set right time 
    try {

      // needed because can't use double quote in ABB/PLC software 
      const rightApexChunk = chunk.toString().replace(/'/g, "\"");
      const data = JSON.parse(rightApexChunk, (key, value) => {
        if(key === "t" && value === "") {
          // TODO - Check the TIMESTAMP value for BigQuery field 
          return Math.floor(Date.now() / 1000)
        } else {
          return value;
        }
      });
      console.log(`@DATA(${socket.remoteAddress}:${socket.remotePort}): ${JSON.stringify(data)}`);

      publishBatchedMessages(JSON.stringify(data))
        .then(() => {
          socket.write('OK');
        })
        .catch(error => {
          // ERROR in Publish - CONSOLE
          const msg = `@ERROR(PUBLISH): ${error}`;
          console.log(msg);
          logError(msg);
          socket.write('ERROR-PUBSUB');
        })

    } catch(e) {
      // ERROR in JSON parsing - CONSOLE 
      const msg = `@ERROR(JSON): ${e}`;
      console.log(msg);
      logError(msg);
      socket.write('ERROR-JSON)');
    }

   });

  // Client request to END the TCP connection, server END the connection
  socket.on('end', () => {
    console.log(`@CLOSING_CONNECTION: ${socket.remoteAddress}:${socket.remotePort}`);
  });

  // Don't forget to catch error, for your own sake.
  socket.on('error', (err) => {
    // ERROR on Socket - CONSOLE 
    const msg = `@ERROR(SOCKET): ${socket.remoteAddress}:${socket.remotePort} - ${err}`;
    console.log(msg);
    logError(msg);
  })
 });

 // -----------------------------------------------------------------------------------------------
